### PR TITLE
ConnCallback

### PR DESCRIPTION
**Pull request by progrium**
This is a relatively quick pass at a new callback that lets you wrap the underlying connection used by SSH, as well as do any other initial connection validation before processing, like determining connection should be dropped for rate limiting.

It also has a `WrapConn` option helper, just because it seems consistent, but it makes me wonder, do we need a helper for every field of server configuration?

### CODE REVIEW

**Comment by progrium** [comment]
Also yes, the test has a race. Not entirely sure how to do that test without clean shutdown.

**Comment by progrium** [comment]
Need to move forward with this. Does this seem ok?

File: server.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ type Server struct {
 	PasswordHandler             PasswordHandler             // password authentication handler
 	PublicKeyHandler            PublicKeyHandler            // public key authentication handler
 	PtyCallback                 PtyCallback                 // callback for allowing PTY sessions, allows all if nil
+  ConnCallback                ConnCallback                // optional callback for wrapping net.Conn before handling
```
Comment (belak) [inline server.go]: Just noticed this... looks like it's indented with spaces? A gofmt might be prudent.

**Comment by progrium** [comment]
Any suggestions to fix the data race in the test?

**Comment by belak** [comment]
You could send "written" on a buffered channel and do a read on that channel with a timeout... then you know it won't hang, but it shouldn't be a race either. Alternatively, `atomic.Value` may be an option.

**Comment by progrium** [comment]
Going to add an example using this and then merge.

**Pull request merged**